juju/juju
# Feature/collapse app config

This is pretty much a forklift copy paste of the existing code with some modifications to make the UI analogous to the model-config and model-defaults commands.

Also, this is not ready to land until the LP issue referenced below is fixed.

Refs: http://pad.lv/1621576
Refs: spec https://goo.gl/yqrrPI

QA:
1. Run all unit tests and they all pass.
2. juju bootstrap reed/aws/us-west-1 --build-agent # bootstrap a controller with settings similar to http://paste.ubuntu.com/23080517/
3. juju deploy haproxy && juju deploy wordpress && juju deploy mysql && juju add-relation haproxy wordpress && juju add-relation wordpress mysql && juju expose haproxy
4. juju config wordpress # verify output is like: http://paste.ubuntu.com/23156363/
5. juju config wordpress debug=yes
6. juju config wordpress debug # verify it is yes
7. juju config wordpress --reset debug
8. save the contents of the mysql settings to a file with `juju config mysql --format=yaml > mysql.yaml 
9. It should look like http://paste.ubuntu.com/23156375/ 
10. Edit file like http://paste.ubuntu.com/23156406/
11.  juju config mysql --file mysql.yaml ## verify the settings are there "ON" and 1.2345679e+08
12. juju config mysql --reset query-cache-size query-cache-type
13. juju config mysql query-cache-size=123456789 query-cache-type=DEMAND
14. juju config mysql ## verify the settings 
15. juju config mysql query-cache-size ## verify a single setting -> 1.2345679e+08 
16. juju config mysql --reset query-cache-size # reset a single 
17. juju config mysql --reset query-cache-size #  -> 0
18  juju config mysql query-cache-type # -> DEMAND
19. juju config mysql --reset query-cache-type
20. juju config mysql --reset # -> error: no configuration options specified
21. uju config mysql foo bar # -> error: can only retrieve a single value, or all values
22. juju destroy-controller reed/aws/us-west-1 -y --destroy-all-models

(Review request: http://reviews.vapour.ws/r/5640/)
